source/fe/mapping_cartesian.cc

```diff
@@ -601,6 +601,90 @@ MappingCartesian<dim, spacedim>::fill_fe_subface_values(
 
 
 
+template <int dim, int spacedim>
+void
+MappingCartesian<dim, spacedim>::fill_fe_immersed_surface_values(
+  const typename Triangulation<dim, spacedim>::cell_iterator &cell,
+  const NonMatching::ImmersedSurfaceQuadrature<dim> &         quadrature,
+  const typename Mapping<dim, spacedim>::InternalDataBase &   internal_data,
+  dealii::internal::FEValuesImplementation::MappingRelatedData<dim, spacedim>
+    &output_data) const
+{
+  Assert(dim == spacedim, ExcNotImplemented());
+
+  // Convert data object to internal data for this class. Fails with an
+  // exception if that is not possible.
+  Assert(dynamic_cast<const InternalData *>(&internal_data) != nullptr,
+         ExcInternalError());
+  const InternalData &data = static_cast<const InternalData &>(internal_data);
```

Comment:
```suggestion
  const auto data = dynamic_cast<const InternalData *>(&internal_data);
  Assert(data != nullptr, ExcInternalError());
```

Reply:
The original code is the same as in all other cases, and its goal is to avoid the (expensive) dynamic cast with its lookup in release mode. This might seem nit-picking but benchmarking things around 12 years ago I realized it actually pays off. :smile:

Reply:
I mean why don't we create a deal.II-macro for this purpose. Dynamic casting is quite common in the library!